docs (external-data): Update readme to include new /register-session-url and /broadcast-signal endpoints (#14652)
microsoft/FluidFramework: examples/hosts/app-integration/external-data/README.md

Reviewer: Nit: the second image for the "Data changes on External Service" has a few typos in step 2.

The diff in Github isn't clear about which images are the final ones in the first flow :(

Many services that would hold the "source of truth" data offer explicit commit style interfaces (e.g. via REST call or similar) which are not well suited to rapid updates.
However, they often expose third-party integration via REST APIS for querying and manipulating data, as well as webhooks for watching updates to the data.

This repo contains a service that mocks the external "source of truth" data server. This mock service offers a REST API collection and webhook interfaces in `./src/mock-external-data-service`. The API requests served by this "external" service are the following:

1. POST `/register-for-webhook`: Register's the sender's URL to receive notifications when the external task-list data changes. Currently, the Customer Service registers its `/external-data-webhook` endpoint here to be called when data changes.
2. GET `/fetch-tasks`: Fetches the task list from the external data store. Called by the Fluid client.
3. POST `/set-tasks`: Updates external data store with new tasks list (complete override). Called by the Fluid client.

> 4. POST `/debug-reset-task-list`: Resets the external data to its original contents. Called by the Fluid client.

Reviewer: Don't feel too strongly about this last one. Keeping it here and saying it's for debugging purposes seems fine too.

> 1. POST `/register-for-webhook?externalTaskListId=XXX`. Required body parameters: `url` (string). Registers the sender's URL to receive notifications when the external task-list data changes. Currently, the Customer Service registers its `/external-data-webhook` endpoint here to be called when data changes.

Reviewer: Super nitty nit: I know it was me who suggested putting the task list id in the querystring. I think I should have suggested adding it to the body. Oh well, no biggie 😄 .

### Customer Service

Next we need a customer service that functions as the intermediary between the External Data Service and the Fluid Service. This server is responsible for authenticating to the external service on the customer's behalf. It registers to the External Data Service webhooks and listens for incoming changes. It also acts as a translation layer, translating to and from the External Data Service expected format and the Fluid Service's expected format.

In this example, the Customer Service contains the following endpoints:

1.POST `/register-session-url`. Required body parameters: `containerUrl` (string), `externalTaskListId` (string). Creates an entry in the Customer Service of the mapping between the container and the external resource id. It then calls the External Service's `/register-for-webhook` endpoint to call it's own `/external-data-webhook?externalTaskListId=XXX` endpoint (details below), in order to notify the Fluid Service containers subscribed to that externalTaskListId of the corresponding change.

> 2. POST `/external-data-webhook`. Required querystring parameters: `externalTaskListId`(string). Registered to be called by the External Data Service webhook when there's been a change to data upstream. On being called, the Customer Service calls the `/broadcast-signal` endpoint in the Fluid Service. Note that this is a route established by the Customer Service to be used exclusively as a subscription endpoint for the external data service's webhook, and should be considered a private implementation detail to the service. Customers may choose to implement this differently.

Reviewer: Nit, rewording
```suggestion
2. POST `/external-data-webhook`. Required querystring parameters: `externalTaskListId`(string). This gets called by the External Data Service when there's been a change to the data and causes the Customer Service to in turn call the `/broadcast-signal` endpoint in the Fluid Service. Note that this is a route established by the Customer Service to be used exclusively as a subscription endpoint for the external data service's webhook and should be considered a private implementation detail to the service. Customers may choose to implement this differently.
```

Find the details of the API in the [Customer Service README](./src/mock-customer-service/README.md).

### Fluid Service

The `broadcast-signal` endpoint is new and still under construction. We will update it here once it is available to use.

> 1. POST `/broadcast-signal`. Required body parameters: `containerUrl` (string), `externalTaskListId` (string), `taskData`(ITaskData). On receiving this, the driver will broadcast a signal of `SignalType.RuntimeSignal` to the clients to alert them of an upstream change. This endpoint is called by the Customer Service to let the Fluid service know that there has been a change in the data. The body must contain the `conainerUrl` that is composed ot the `socketStreamUrl`, the `containerId` (sometimes known as the documentId), and the tenantId. In this way, it "echoes" the webhook from the External Data Service to the Customer Service. It is called by the Customer Service when it needs to notify the Fluid Clients that there has been a change to the data.

Reviewer: Nit, rewording
```suggestion
1. POST `/broadcast-signal`. Required body parameters: `containerUrl` (string), `externalTaskListId` (string), `taskData`(ITaskData). This endpoint is called by the Customer Service when it needs the Fluid Service to notify the Fluid clients that there has been a change to the upstream data. The body must contain the `containerUrl`, from which we can extract a few pieces relevant to the Fluid service: the `socketStreamUrl`, the `containerId` (sometimes known as the `documentId`), and the `tenantId`. The Fluid service will broadcast a signal of type `SignalType.RuntimeSignal` to the clients to alert them of the change. In this way, it "echoes" the webhook call from the External Data Service to the Customer Service.
```

> This example uses the example tinylicious driver to stub out what changes will be necessary in the odsp-driver. The prototype of the full signal and driver flow can be seen in this [`dev/external-data-prototyping` branch to main comparison](https://github.com/microsoft/FluidFramework/compare/main...dev/external-data-prototyping).

Reviewer: ```suggestion
This example uses the tinylicious driver to stub out what changes will be necessary in the odsp-driver. The prototype of the full signal and driver flow can be seen in this [`dev/external-data-prototyping` branch to main comparison](https://github.com/microsoft/FluidFramework/compare/main...dev/external-data-prototyping).
```

2. POST `/external-data-webhook`. Required querystring parameters: `externalTaskListId`(string). Registered to be called by the External Data Service webhook when there's been a change to data upstream. On being called, the Customer Service calls the `/broadcast-signal` endpoint in the Fluid Service. Note that this is a route established by the Customer Service to be used exclusively as a subscription endpoint for the external data service's webhook, and should be considered a private implementation detail to the service. Customers may choose to implement this differently.

Find the details of the API in the [Customer Service README](./src/mock-customer-service/README.md).

**Fluid Service**

_Driver Layer_

> The driver layer section is internal facing and anyone implementing this pattern need not be concerned with this. The changes in the driver layer are a new (and still under construction) `broadcast-signal` endpoint and the logic to relay the signal to the correct Fluid container when called.

Reviewer 2: This was a bit hard to parse. Take these suggestions with a grain of salt, but they might make it a bit easier to read.
```suggestion
The driver layer section is internal facing, and anyone implementing this pattern need not be concerned with it. This layer is new (and still under construction). It exposes a `broadcast-signal` endpoint and the logic to relay the signal to the correct Fluid container when called.
```

Author: Hmm you know, thinking about it -- it's not true that the implementor doesn't have to be concerned with it since they do need to explicitly call it within their implementation for the customer service. Maybe I should just leave this whole caveatout, except the bit where it's still a bit under construction and might change.

TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.

TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.

ITaskData - This is similar to Task except that it is the External Data Services model of the task.

ITaskListData - This is similar to TaskList except that it is the External Data Services model of the taskList.

> <!-- prettier-ignore-start -->

Reviewer 2: Is there a specific reason we have to ignore prettier here?

Author: It's complaining about the non-capitalized f for fluid in the SharedMap link (https://fluidframework.com/docs/data-structures/map/)

Reviewer: Wait, we do have a markdown linter!? Why does it force us to put four spaces between the dash and the text in a list item and why doesn't it force blank lines before and after headers? 😡

Sorry for the tangent.

Author: Ooo you're right -- okay I was able to remove this!

TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.

> TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.

Reviewer: "All tasks" makes me think of "all the tasks in the JIRA instance".
```suggestion
TaskList - This can be compared to a "board" that holds a list of related tasks; for example, the results of a particular query. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.
```
